osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/messaging/processors/KeysRequestMessageProcessor.java

```diff
@@ -0,0 +1,108 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.processors;
+
+import java.io.Serializable;
+import java.time.Duration;
+import javax.jms.JMSException;
+import javax.jms.ObjectMessage;
+import lombok.extern.slf4j.Slf4j;
+import org.opensmartgridplatform.adapter.protocol.dlms.application.services.ConfigurationService;
+import org.opensmartgridplatform.adapter.protocol.dlms.application.services.DeviceKeyProcessingService;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.factories.DlmsConnectionManager;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.DeviceKeyProcessAlreadyRunningException;
+import org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.DeviceRequestMessageProcessor;
+import org.opensmartgridplatform.adapter.protocol.dlms.infra.messaging.DeviceRequestMessageSender;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.SetKeysRequestDto;
+import org.opensmartgridplatform.shared.exceptionhandling.OsgpException;
+import org.opensmartgridplatform.shared.infra.jms.MessageMetadata;
+import org.opensmartgridplatform.shared.infra.jms.MessageType;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@Slf4j
+public class KeysRequestMessageProcessor extends DeviceRequestMessageProcessor {
+
+  @Autowired private ConfigurationService configurationService;
+
+  @Autowired private DeviceRequestMessageSender deviceRequestMessageSender;
+
+  @Autowired private DeviceKeyProcessingService deviceKeyProcessingService;
+
+  public KeysRequestMessageProcessor(final MessageType messageType) {
+    super(messageType);
+  }
+
+  @Override
+  protected Serializable handleMessage(
+      final DlmsConnectionManager conn,
+      final DlmsDevice device,
+      final Serializable requestObject,
+      final MessageMetadata messageMetadata)
+      throws OsgpException {
+
+    this.assertRequestObjectType(SetKeysRequestDto.class, requestObject);
+
+    final SetKeysRequestDto keySet = (SetKeysRequestDto) requestObject;
+    this.configurationService.replaceKeys(conn, device, keySet, messageMetadata);
+
+    return null;
+  }
+
+  @Override
+  public void processMessage(final ObjectMessage message) throws JMSException {
+    final MessageMetadata messageMetadata = MessageMetadata.fromMessage(message);
+    final Serializable messageObject = message.getObject();
+
+    log.debug(
+        "Checking if not already processing key for device {} before processing {} request message",
+        messageMetadata.getDeviceIdentification(),
+        messageMetadata.getMessageType());
+
+    try {
+      final DlmsDevice device = this.domainHelperService.findDlmsDevice(messageMetadata);
+
+      this.checkIfOtherKeyProcessIsRunning(device.getDeviceIdentification());
+
+      this.markKeyProcessIsRunning(device.getDeviceIdentification());
```

Comment:
This leaves open a small window, where two processes can check if another key process is running, and see that it is not, after which both could go into marking a key process is running, where - because of the way that is implemented - one process will create a new `DeviceKeyProcessing` and the other will find it and update its start time.

I believe a similar risk is introduced in other locations where the check and the mark action are separated (for instance in the `RecoverKeyProcess`).

Reply:
Should there be a finally block indicating the process is no longer running, or was this left out intentionally for some reason?

Reply:
check and mark method are merge into one call to the database. If insert in device_Key_Processing table succeeds then processing is started if it fails an update -on record older than timeout- is fired. If this fails there is a process running if succeeds then processing is started.
stopProcessing is added in finally block and in catch ThrottlingPermitDeniedException block